skepticoin/scripts/utils.py

```diff
@@ -63,57 +63,64 @@ def read_chain_from_disk() -> CoinState:
-                        print(filename)
+    else:
+        try:
+            print("Pre-download blockchain from trusted source to 'chain.zip'")
+            with urllib.request.urlopen(TRUSTED_BLOCKCHAIN_ZIP) as resp:
+                with open('chain.zip', 'wb') as outfile:
+                    outfile.write(resp.read())
+            print("Reading initial chain from zipfile")
+            coinstate = CoinState.zero()
+            with zipfile.ZipFile('chain.zip') as zip:
+                for entry in zip.infolist():
+                    if not entry.is_dir():
+                        filename = entry.filename.split('/')[1]
+                        height = int(filename.split("-")[0])
+                        if height % 1000 == 0:
+                            print(filename)
 
-                    try:
                         data = zip.read(entry)
                         block = Block.stream_deserialize(BytesIO(data))
-                    except Exception as e:
-                        raise Exception("Corrupted block on disk: %s" % filename) from e
+                        coinstate = coinstate.add_block_no_validation(block)
+        except Exception:
+            print("Error reading zip file. We'll start with an empty blockchain instead." + traceback.format_exc())
+            coinstate = CoinState.zero()
 
-                    coinstate = coinstate.add_block_no_validation(block)
+    rewrite = False
 
-    else:
-        coinstate = CoinState.zero()
-        height = 0
+    if os.path.isdir('chain'):
+        # the code below is no longer needed by normal users, but some old testcases still rely on it:
+        for filename in sorted(os.listdir('chain')):
+            (height_str, hash_str) = filename.split("-")
+            (height, hash) = (int(height_str), computer(hash_str))
 
-    fresher_chain = sorted(os.listdir('chain'))[height:]
-    print("Reading chain from disk, starting height=%d, fresher=%d" % (height, len(fresher_chain)))
-    for filename in fresher_chain:
-        height = int(filename.split("-")[0])
-        if height % 1000 == 0:
-            print(filename)
+            if hash not in coinstate.block_by_hash:
 
-        try:
-            with open(Path("chain") / filename, 'rb') as f:
-                block = Block.stream_deserialize(f)
-        except Exception as e:
-            raise Exception("Corrupted block on disk: %s" % filename) from e
+                if height % 1000 == 0:
+                    print(filename)
 
-        coinstate = coinstate.add_block_no_validation(block)
+                if os.path.getsize(f"chain/{filename}") == 0:
+                    print("Stopping at empty block file: %s" % filename)
+                    break
 
-    if fresher_chain:
-        write_chain_cache_to_disk(coinstate)
+                with open(Path("chain") / filename, 'rb') as f:
+                    try:
+                        block = Block.stream_deserialize(f)
+                    except Exception as e:
```

Comment:
Time to implement some proper exception handling, see #91. Not in this PR of course.

Reply:
I agree!